src/mbgl/text/placement.cpp

```diff
@@ -143,419 +230,388 @@ Point<float> calculateVariableLayoutOffset(style::SymbolAnchorType anchor,
-    const bool textAllowOverlap = layout.get<style::TextAllowOverlap>();
-    const bool iconAllowOverlap = layout.get<style::IconAllowOverlap>();
-    // This logic is similar to the "defaultOpacityState" logic below in updateBucketOpacities
-    // If we know a symbol is always supposed to show, force it to be marked visible even if
-    // it wasn't placed into the collision index (because some or all of it was outside the range
-    // of the collision grid).
-    // There is a subtle edge case here we're accepting:
-    //  Symbol A has text-allow-overlap: true, icon-allow-overlap: true, icon-optional: false
-    //  A's icon is outside the grid, so doesn't get placed
-    //  A's text would be inside grid, but doesn't get placed because of icon-optional: false
-    //  We still show A because of the allow-overlap settings.
-    //  Symbol B has allow-overlap: false, and gets placed where A's text would be
-    //  On panning in, there is a short period when Symbol B and Symbol A will overlap
-    //  This is the reverse of our normal policy of "fade in on pan", but should look like any other
-    //  collision and hopefully not be too noticeable.
-    // See https://github.com/mapbox/mapbox-gl-native/issues/12683
-    const bool alwaysShowText = textAllowOverlap && (iconAllowOverlap || !(bucket.hasIconData() || bucket.hasSdfIconData()) || layout.get<style::IconOptional>());
-    const bool alwaysShowIcon = iconAllowOverlap && (textAllowOverlap || !bucket.hasTextData() || layout.get<style::TextOptional>());
-    std::vector<style::TextVariableAnchorType> variableTextAnchors = layout.get<style::TextVariableAnchor>();
+        const auto placeTextForPlacementModes = [&](auto& placeHorizontalFn, auto& placeVerticalFn) {
+            if (bucket.allowVerticalPlacement && symbolInstance.writingModes & WritingModeType::Vertical) {
+                assert(!bucket.placementModes.empty());
+                for (auto& placementMode : bucket.placementModes) {
+                    if (placementMode == TextWritingModeType::Vertical) {
+                        placedVerticalText = placed = placeVerticalFn();
+                    } else {
+                        placed = placeHorizontalFn();
+                    }
 
-    const bool hasIconTextFit = layout.get<style::IconTextFit>() != style::IconTextFitType::None;
+                    if (placed.first) {
+                        break;
+                    }
+                }
+            } else {
+                placed = placeHorizontalFn();
+            }
+        };
 
-    std::vector<ProjectedCollisionBox> textBoxes;
-    std::vector<ProjectedCollisionBox> iconBoxes;
+        // Line or point label placement
+        if (variableTextAnchors.empty()) {
+            const auto placeFeature = [&](const CollisionFeature& collisionFeature,
+                                          style::TextWritingModeType orientation) {
+                textBoxes.clear();
```

Comment:
They do we clear placement cache again here and also one more time down this code?

Reply:
yeah, one time is enough :smile:

Reply:
actually yes, and also at each variable placement attempt